dali/kernels/signal/dct/dct_gpu_test.cc

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <tuple>
+#include "dali/kernels/signal/dct/dct_gpu.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/test/tensor_test_utils.h"
+#include "dali/test/test_tensors.h"
+#include "dali/kernels/signal/dct/dct_test.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+namespace dct {
+namespace test {
+
+constexpr int Dims = 4;
+
+class Dct1DGpuTest : public ::testing::TestWithParam<std::tuple<int, std::vector<int>>> {
+ public:
+  Dct1DGpuTest()
+      : batch_size_(std::get<0>(GetParam()))
+      , axes_(std::get<1>(GetParam()))
+      , in_shape_(batch_size_) {}
+
+  ~Dct1DGpuTest() override = default;
+
+ protected:
+  void PrepareInput() {
+    args_.clear();
+    std::mt19937_64 rng{1223};
+    std::uniform_int_distribution<> dim_dist(1, 3);
+    auto rand_shape = [&]() {
+      TensorShape<Dims> shape;
+      for (int i = 0; i < Dims; ++i)
+        shape[i] = dim_dist(rng) * 10;
+      return shape;
+    };
+    for (int s = 0; s < batch_size_; ++s) {
+      auto i = arg_idx_;
+      in_shape_.set_tensor_shape(s, rand_shape());
+      args_.emplace_back(get_args());
+    }
+    ttl_in_.reshape(in_shape_);
+    UniformRandomFill(ttl_in_.cpu(), rng, 0., 1.);
+  }
+
+  TensorListShape<Dims> OutputShape(int axis) {
+    TensorListShape<Dims> out_shape(batch_size_);
+    for (int s = 0; s < batch_size_; ++s) {
+      auto in_shape = in_shape_[s];
+      int ndct = args_[s].ndct;
+      if (ndct > 0)
+        in_shape[axis] = ndct;
+      out_shape.set_tensor_shape(s, in_shape);
+    }
+    return out_shape;
+  }
+
+  DctArgs get_args() {
+    auto i = arg_idx_;
+    if (normalize[(i / 3) % 2] && dct_type[(i / 6) % 4] == 1) {
```

Comment:
Can you put a comment where this math comes from?

Reply:
I've changed the way I iterate over the cartesian product of args, so it's no longer here